Let the compiler generate more comparison operators in WebCore

cdumez opened the pull request:
#### f61c5111898cf9fd9ee106240f085c69776dc8a8
<pre>
Let the compiler generate more comparison operators in WebCore
<a href="https://bugs.webkit.org/show_bug.cgi?id=261012">https://bugs.webkit.org/show_bug.cgi?id=261012</a>

Reviewed by Darin Adler.

Let the compiler generate more comparison operators in WebCore now that we
support C++20.

* Source/WebCore/css/CSSProperty.h:
(WebCore::StylePropertyMetadata::operator== const): Deleted.
* Source/WebCore/css/StyleColor.h:
* Source/WebCore/css/color/CSSUnresolvedColorMix.cpp:
(WebCore::operator==):
* Source/WebCore/css/color/CSSUnresolvedColorMix.h:
* Source/WebCore/css/parser/CSSParserContext.cpp:
(WebCore::operator==): Deleted.
* Source/WebCore/css/parser/CSSParserContext.h:
* Source/WebCore/css/parser/CSSSelectorParser.h:
* Source/WebCore/css/typedom/numeric/CSSNumericType.cpp:
(WebCore::CSSNumericType::operator== const): Deleted.
* Source/WebCore/css/typedom/numeric/CSSNumericType.h:
* Source/WebCore/dom/ComposedTreeAncestorIterator.h:
(WebCore::ComposedTreeAncestorIterator::operator== const): Deleted.
* Source/WebCore/dom/MessagePortIdentifier.h:
(WebCore::operator==): Deleted.
* Source/WebCore/dom/QualifiedName.h:
(WebCore::QualifiedName::operator== const): Deleted.
* Source/WebCore/dom/SimpleRange.cpp:
(WebCore::operator==): Deleted.
* Source/WebCore/dom/SimpleRange.h:
* Source/WebCore/dom/SpaceSplitString.h:
(WebCore::SpaceSplitString::operator== const): Deleted.
* Source/WebCore/editing/cocoa/EditorCocoa.mm:
(WebCore::Editor::replaceNodeFromPasteboard):
* Source/WebCore/html/HTMLElement.cpp:
(WebCore::checkPopoverValidity):
* Source/WebCore/html/LinkRelAttribute.h:
(WebCore::operator==): Deleted.
* Source/WebCore/html/canvas/WebGLFramebuffer.cpp:
(WebCore::WebGLFramebuffer::TextureAttachment::operator== const): Deleted.
(WebCore::WebGLFramebuffer::TextureLayerAttachment::operator== const): Deleted.
* Source/WebCore/html/canvas/WebGLFramebuffer.h:
* Source/WebCore/inspector/agents/InspectorNetworkAgent.h:
(WebCore::InspectorNetworkAgent::Intercept::operator== const): Deleted.
* Source/WebCore/layout/LayoutUnits.h:
(WebCore::Layout::operator==): Deleted.
* Source/WebCore/layout/formattingContexts/inline/InlineLevelBox.h:
(WebCore::Layout::InlineLevelBox::AscentAndDescent::operator== const): Deleted.
* Source/WebCore/layout/formattingContexts/inline/InlineLineTypes.h:
(WebCore::Layout::InlineItemPosition::operator== const): Deleted.
* Source/WebCore/layout/integration/inline/InlineIteratorBoxLegacyPath.h:
(WebCore::InlineIterator::BoxLegacyPath::operator== const): Deleted.
* Source/WebCore/layout/integration/inline/InlineIteratorBoxModernPath.h:
(WebCore::InlineIterator::BoxModernPath::operator== const): Deleted.
* Source/WebCore/layout/integration/inline/InlineIteratorLineBoxLegacyPath.h:
(WebCore::InlineIterator::LineBoxIteratorLegacyPath::operator== const): Deleted.
* Source/WebCore/layout/integration/inline/InlineIteratorLineBoxModernPath.h:
(WebCore::InlineIterator::LineBoxIteratorModernPath::operator== const): Deleted.
* Source/WebCore/layout/layouttree/LayoutContainingBlockChainIterator.h:
(WebCore::Layout::LayoutContainingBlockChainIterator::operator== const): Deleted.
* Source/WebCore/loader/CrossOriginEmbedderPolicy.h:
(WebCore::operator==): Deleted.
* Source/WebCore/loader/CrossOriginOpenerPolicy.h:
(WebCore::operator==): Deleted.
* Source/WebCore/loader/PCMSites.h:
(WebCore::PCM::SourceSite::operator== const): Deleted.
(WebCore::PCM::AttributionDestinationSite::operator== const): Deleted.
* Source/WebCore/loader/PolicyContainer.h:
(WebCore::operator==): Deleted.
* Source/WebCore/loader/ResourceCryptographicDigest.h:
(WebCore::ResourceCryptographicDigest::operator== const): Deleted.
* Source/WebCore/page/ClientOrigin.h:
(WebCore::ClientOrigin::operator== const): Deleted.
* Source/WebCore/page/GlobalWindowIdentifier.h:
(WebCore::operator==): Deleted.
* Source/WebCore/page/InteractionRegion.h:
(WebCore::operator==): Deleted.
* Source/WebCore/page/LocalFrameView.h:
* Source/WebCore/page/SecurityOriginData.h:
(WebCore::SecurityOriginData::Tuple::operator== const): Deleted.
* Source/WebCore/page/ViewportConfiguration.h:
(WebCore::ViewportConfiguration::Parameters::operator== const): Deleted.
* Source/WebCore/page/csp/ContentSecurityPolicy.cpp:
(WebCore::ContentSecurityPolicy::reportViolation const):
* Source/WebCore/page/scrolling/ScrollingConstraints.h:
(WebCore::AbsolutePositionConstraints::operator== const): Deleted.
(WebCore::FixedPositionViewportConstraints::operator== const): Deleted.
(WebCore::StickyPositionViewportConstraints::operator== const): Deleted.
* Source/WebCore/page/scrolling/ScrollingCoordinatorTypes.h:
(WebCore::ScrollableAreaParameters::operator== const): Deleted.
(WebCore::RequestedKeyboardScrollData::operator== const): Deleted.
* Source/WebCore/page/scrolling/ScrollingStateScrollingNode.h:
(WebCore::ScrollbarHoverState::operator== const): Deleted.
* Source/WebCore/platform/Decimal.h:
* Source/WebCore/platform/EventTrackingRegions.cpp:
(WebCore::operator==): Deleted.
* Source/WebCore/platform/EventTrackingRegions.h:
* Source/WebCore/platform/KeyboardScroll.h:
(WebCore::KeyboardScroll::operator== const): Deleted.
* Source/WebCore/platform/LayoutUnit.h:
* Source/WebCore/platform/LengthPoint.h:
(WebCore::LengthPoint::operator== const): Deleted.
* Source/WebCore/platform/LengthSize.h:
(WebCore::operator==): Deleted.
* Source/WebCore/platform/ProcessQualified.h:
(WebCore::ProcessQualified::operator== const): Deleted.
* Source/WebCore/platform/RectEdges.h:
(WebCore::RectEdges::operator== const): Deleted.
* Source/WebCore/platform/RegistrableDomain.h:
* Source/WebCore/platform/ScriptExecutionContextIdentifier.h:
(WebCore::ProcessQualified&lt;WTF::UUID&gt;::operator== const): Deleted.
* Source/WebCore/platform/Timer.cpp:
(WebCore::operator==): Deleted.
* Source/WebCore/platform/audio/NowPlayingInfo.h:
(WebCore::NowPlayingInfo::operator== const): Deleted.

Canonical link: <a href="https://commits.webkit.org/267580@main">https://commits.webkit.org/267580@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/20186b3eeec36a94b7df40e40e320d067c4e11b7

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/17053 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/17376 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/17851 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/18838 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/15953 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/17243 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/20634 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/17517 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/18170 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/17254 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/17617 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/14788 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/19653 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/14857 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/15479 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/22184 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/15850 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/15646 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/19999 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/16239 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/13768 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/15398 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4072 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/19761 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/16072 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/editing/cocoa/EditorCocoa.mm:
```diff
@@ -342,7 +342,7 @@ static void maybeCopyNodeAttributesToFragment(const Node& node, DocumentFragment
 void Editor::replaceNodeFromPasteboard(Node& node, const String& pasteboardName, EditAction action)
 {
-    if (node.document() != m_document)
+    if (&node.document() != &m_document)
```
Does this fix a bug? What did the code end up doing before?

cdumez replied:
Ah yes, I was a bit surprised the previous code was building. I am not sure what it ended up doing. The `CSSParserContext` constructor that takes in a `Document&` is not marked as explicit and I modified CSSParserContext's operator==() in this patch so maybe it was using that?

I can double check.

cdumez replied:
Yes, I was right. I was using CSSParserContext's operator==() :/
I will write a patch to make CSSParserContext's constructors explicit.

darinadler replied:
Amazing! That is so nuts. That may even have been costly enough that we get a tiny performance boost.

cdumez replied:
will land https://github.com/WebKit/WebKit/pull/17356/ first to avoid behavior changes in this large refactoring patch.

darinadler commented on Source/WebCore/html/HTMLElement.cpp:
```diff
@@ -1249,7 +1249,7 @@ static ExceptionOr<bool> checkPopoverValidity(HTMLElement& element, PopoverVisib
         return Exception { InvalidStateError, "Element is not connected"_s };
 
-    if (expectedDocument && element.document() != *expectedDocument)
+    if (expectedDocument && &element.document() != expectedDocument)
```
Does this fix a bug? What did the code end up doing before?

darinadler commented on Source/WebCore/layout/LayoutUnits.h:
```diff
@@ -52,6 +52,8 @@ using InlineLayoutRect = LayoutRect;
 
 struct Position {
     operator LayoutUnit() const { return value; }
+    friend bool operator==(const Position&, const Position&) = default;
```
I can’t figure out what rule you are using to decide when to use `const&` and when to just pass the structure. I would have expected just `Position` here, but I’m not sure it matters.

cdumez replied:
I'll switch to `Position`, I didn't realize the type was so small.

darinadler replied:
There are others like this even in this patch, such as `LayoutUnit` itself. I didn’t try to thoroughly search for such things.

darinadler commented on Source/WebCore/page/csp/ContentSecurityPolicy.cpp:
```diff
@@ -894,7 +894,7 @@ void ContentSecurityPolicy::reportViolation(const String& effectiveViolatedDirec
     else {
         auto& document = downcast<Document>(*m_scriptExecutionContext);
-        if (element && element->document() == document)
+        if (element && &element->document() == &document)
```
Does this fix a bug? What did the code end up doing before?

darinadler commented on Source/WebCore/platform/Timer.cpp:
```diff
@@ -183,7 +183,7 @@ class TimerHeapIterator {
     }
 
-    friend bool operator==(TimerHeapIterator, TimerHeapIterator);
+    friend bool operator==(TimerHeapIterator, TimerHeapIterator) = default;
```
I think we might, in the future, want to generate `<=>` instead of `==` here.

cdumez committed:
Let the compiler generate more comparison operators in WebCore
https://bugs.webkit.org/show_bug.cgi?id=261012

Reviewed by Darin Adler.

Let the compiler generate more comparison operators in WebCore now that we
support C++20.

* Source/WebCore/css/CSSProperty.h:
(WebCore::StylePropertyMetadata::operator== const): Deleted.
* Source/WebCore/css/StyleColor.h:
* Source/WebCore/css/color/CSSUnresolvedColorMix.cpp:
(WebCore::operator==):
* Source/WebCore/css/color/CSSUnresolvedColorMix.h:
* Source/WebCore/css/parser/CSSParserContext.cpp:
(WebCore::operator==): Deleted.
* Source/WebCore/css/parser/CSSParserContext.h:
* Source/WebCore/css/parser/CSSSelectorParser.h:
* Source/WebCore/css/typedom/numeric/CSSNumericType.cpp:
(WebCore::CSSNumericType::operator== const): Deleted.
* Source/WebCore/css/typedom/numeric/CSSNumericType.h:
* Source/WebCore/dom/ComposedTreeAncestorIterator.h:
(WebCore::ComposedTreeAncestorIterator::operator== const): Deleted.
* Source/WebCore/dom/MessagePortIdentifier.h:
(WebCore::operator==): Deleted.
* Source/WebCore/dom/QualifiedName.h:
(WebCore::QualifiedName::operator== const): Deleted.
* Source/WebCore/dom/SimpleRange.cpp:
(WebCore::operator==): Deleted.
* Source/WebCore/dom/SimpleRange.h:
* Source/WebCore/dom/SpaceSplitString.h:
(WebCore::SpaceSplitString::operator== const): Deleted.
* Source/WebCore/editing/cocoa/EditorCocoa.mm:
(WebCore::Editor::replaceNodeFromPasteboard):
* Source/WebCore/html/HTMLElement.cpp:
(WebCore::checkPopoverValidity):
* Source/WebCore/html/LinkRelAttribute.h:
(WebCore::operator==): Deleted.
* Source/WebCore/html/canvas/WebGLFramebuffer.cpp:
(WebCore::WebGLFramebuffer::TextureAttachment::operator== const): Deleted.
(WebCore::WebGLFramebuffer::TextureLayerAttachment::operator== const): Deleted.
* Source/WebCore/html/canvas/WebGLFramebuffer.h:
* Source/WebCore/inspector/agents/InspectorNetworkAgent.h:
(WebCore::InspectorNetworkAgent::Intercept::operator== const): Deleted.
* Source/WebCore/layout/LayoutUnits.h:
(WebCore::Layout::operator==): Deleted.
* Source/WebCore/layout/formattingContexts/inline/InlineLevelBox.h:
(WebCore::Layout::InlineLevelBox::AscentAndDescent::operator== const): Deleted.
* Source/WebCore/layout/formattingContexts/inline/InlineLineTypes.h:
(WebCore::Layout::InlineItemPosition::operator== const): Deleted.
* Source/WebCore/layout/integration/inline/InlineIteratorBoxLegacyPath.h:
(WebCore::InlineIterator::BoxLegacyPath::operator== const): Deleted.
* Source/WebCore/layout/integration/inline/InlineIteratorBoxModernPath.h:
(WebCore::InlineIterator::BoxModernPath::operator== const): Deleted.
* Source/WebCore/layout/integration/inline/InlineIteratorLineBoxLegacyPath.h:
(WebCore::InlineIterator::LineBoxIteratorLegacyPath::operator== const): Deleted.
* Source/WebCore/layout/integration/inline/InlineIteratorLineBoxModernPath.h:
(WebCore::InlineIterator::LineBoxIteratorModernPath::operator== const): Deleted.
* Source/WebCore/layout/layouttree/LayoutContainingBlockChainIterator.h:
(WebCore::Layout::LayoutContainingBlockChainIterator::operator== const): Deleted.
* Source/WebCore/loader/CrossOriginEmbedderPolicy.h:
(WebCore::operator==): Deleted.
* Source/WebCore/loader/CrossOriginOpenerPolicy.h:
(WebCore::operator==): Deleted.
* Source/WebCore/loader/PCMSites.h:
(WebCore::PCM::SourceSite::operator== const): Deleted.
(WebCore::PCM::AttributionDestinationSite::operator== const): Deleted.
* Source/WebCore/loader/PolicyContainer.h:
(WebCore::operator==): Deleted.
* Source/WebCore/loader/ResourceCryptographicDigest.h:
(WebCore::ResourceCryptographicDigest::operator== const): Deleted.
* Source/WebCore/page/ClientOrigin.h:
(WebCore::ClientOrigin::operator== const): Deleted.
* Source/WebCore/page/GlobalWindowIdentifier.h:
(WebCore::operator==): Deleted.
* Source/WebCore/page/InteractionRegion.h:
(WebCore::operator==): Deleted.
* Source/WebCore/page/LocalFrameView.h:
* Source/WebCore/page/SecurityOriginData.h:
(WebCore::SecurityOriginData::Tuple::operator== const): Deleted.
* Source/WebCore/page/ViewportConfiguration.h:
(WebCore::ViewportConfiguration::Parameters::operator== const): Deleted.
* Source/WebCore/page/csp/ContentSecurityPolicy.cpp:
(WebCore::ContentSecurityPolicy::reportViolation const):
* Source/WebCore/page/scrolling/ScrollingConstraints.h:
(WebCore::AbsolutePositionConstraints::operator== const): Deleted.
(WebCore::FixedPositionViewportConstraints::operator== const): Deleted.
(WebCore::StickyPositionViewportConstraints::operator== const): Deleted.
* Source/WebCore/page/scrolling/ScrollingCoordinatorTypes.h:
(WebCore::ScrollableAreaParameters::operator== const): Deleted.
(WebCore::RequestedKeyboardScrollData::operator== const): Deleted.
* Source/WebCore/page/scrolling/ScrollingStateScrollingNode.h:
(WebCore::ScrollbarHoverState::operator== const): Deleted.
* Source/WebCore/platform/Decimal.h:
* Source/WebCore/platform/EventTrackingRegions.cpp:
(WebCore::operator==): Deleted.
* Source/WebCore/platform/EventTrackingRegions.h:
* Source/WebCore/platform/KeyboardScroll.h:
(WebCore::KeyboardScroll::operator== const): Deleted.
* Source/WebCore/platform/LayoutUnit.h:
* Source/WebCore/platform/LengthPoint.h:
(WebCore::LengthPoint::operator== const): Deleted.
* Source/WebCore/platform/LengthSize.h:
(WebCore::operator==): Deleted.
* Source/WebCore/platform/ProcessQualified.h:
(WebCore::ProcessQualified::operator== const): Deleted.
* Source/WebCore/platform/RectEdges.h:
(WebCore::RectEdges::operator== const): Deleted.
* Source/WebCore/platform/RegistrableDomain.h:
* Source/WebCore/platform/ScriptExecutionContextIdentifier.h:
(WebCore::ProcessQualified<WTF::UUID>::operator== const): Deleted.
* Source/WebCore/platform/Timer.cpp:
(WebCore::operator==): Deleted.
* Source/WebCore/platform/audio/NowPlayingInfo.h:
(WebCore::NowPlayingInfo::operator== const): Deleted.

Canonical link: https://commits.webkit.org/267580@main
